tools/restatectl/src/commands/node/disable_node_checker.rs

```diff
@@ -0,0 +1,121 @@
+            .map_err(|_err| DisableNodeError::UnknownNode(node_id))?;
+
+        self.safe_to_disable_log_server(node_id, node_config.log_server_config.storage_state)?;
+
+        // only safe to disable node if it does not run a metadata server or is not a member
+        // todo atm we must consider the role because the default metadata server state is MetadataServerState::Member.
+        //  We need to introduce a provisioning state or make the metadata server state optional in the NodeConfig.
+        if node_config.roles.contains(Role::MetadataServer)
+            && node_config.metadata_server_config.metadata_server_state
+                == MetadataServerState::Member
+        {
+            return Err(DisableNodeError::MetadataMember);
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether it is safe to disable the given log server identified by the node_id. It is
+    /// safe to disable the log server if it can no longer be added to new node sets (== not being
+    /// a candidate) and it is no longer part of any known node sets.
+    fn safe_to_disable_log_server(
+        &self,
+        node_id: PlainNodeId,
+        storage_state: StorageState,
+    ) -> Result<(), DisableNodeError> {
+        match storage_state {
+            StorageState::ReadWrite => {
+                return Err(DisableNodeError::ReadWrite);
+            }
+            // it's safe to disable a disabled node
+            StorageState::Disabled => return Ok(()),
+            // we need to check whether this node is no longer part of any known node sets
+            StorageState::ReadOnly | StorageState::DataLoss | StorageState::Provisioning => {}
+        }
+
+        // if the default provider kind is local or in-memory than it is not safe to disable the
+        // given node because it is included in the implicit node set
+        if matches!(
+            self.logs.configuration().default_provider.kind(),
+            ProviderKind::Local | ProviderKind::InMemory
+        ) {
+            return Err(DisableNodeError::DefaultLogletProvider);
+        }
+
+        // check for every log that the given node is not contained in any node set
+        for (log_id, chain) in self.logs.iter() {
+            for segment in chain.iter().rev() {
+                if match segment.config.kind {
+                    // we assume that the given node runs the local and memory loglet and, therefore,
+                    // is part of the implicit node set
+                    ProviderKind::Local => true,
+                    ProviderKind::InMemory => true,
+                    ProviderKind::Replicated => {
+                        let replicated_loglet = ReplicatedLogletParams::deserialize_from(
+                            segment.config.params.as_bytes(),
+                        )
+                        .expect("to be deserializable");
+                        replicated_loglet.nodeset.contains(node_id)
+                    }
```

Comment:
Perhaps you can have a riff of this version to leverage the loglets cache instead of deserializing on demand.
```rs        
let count_loglets = logs
            .iter_replicated_loglets()
            .filter(|(_, loglet)| loglet.params.nodeset.contains(node_id))
            .count();
```

Reply:
Thanks for pointing this out to me. I totally forgot about the `LookupIndex`.